src/routing/approved.rs

```diff
@@ -1005,14 +1019,33 @@ impl Approved {
                 );
                 return Ok(vec![]);
             } else {
-                (MIN_AGE, None, None)
+                // Start as Adult as long as passed resource proofing.
+                (MIN_AGE + 1, None, None)
             };
 
-        self.vote(Vote::Online {
-            member_info: MemberInfo::joined(peer.with_age(age)),
-            previous_name,
-            their_knowledge,
-        })
+        if let Some((solution, data)) = join_request.resource_proof {
+            if let Some(nonce) = self.nonces.get(peer.name()) {
+                if self.resource_proof.validate_all(nonce, &data, solution) {
+                    return self.vote(Vote::Online {
+                        member_info: MemberInfo::joined(peer.with_age(age)),
+                        previous_name,
+                        their_knowledge,
+                    });
+                }
+            }
+        }
+
+        let nonce: [u8; 32] = rand::random();
+        let _ = self.nonces.insert(*peer.name(), nonce);
+        let response = BootstrapResponse::ResourceProof {
```

Comment:
Two (minor) issues with this:
1. I don't think this should be a `BootstrapResponse` variant, because we use bootstrap request/response just to find the section to join. I'd make it a separate message variant.
2. The name "resource poof" is overloaded in a confusing way. It sometimes means the request/challenge and sometimes the response/solution. We should use distinct names for these two concepts.

Reply:
> I'd make it a separate message variant.
> We should use distinct names for these two concepts.

How about `Variant::Challenge` for request/challenge . and `proof` for response/solution ?

Reply:
Agree, but I'd make it `Variant::ResourceChallenge` to be more explicit.